### PR TITLE
Discontinuous density BCs

### PR DESCRIPTION
**Pull request by valerian-chen**
Hey @ZedThree, I'm implementing the generalised BCs for the plasma-vacuum boundary. I've attached a rough summary of the equations. Bolded text refers to things I'm not sure about. Also, not all the prose might be completely updated, but the equations should be good.
[main.pdf](https://github.com/beam-tracing/Scotty/files/11189298/main.pdf)

### CODE REVIEW

**Comment by valerian-chen** [comment]
Note to self: ask whether we can get the electron density directly from the Hamiltonian object

**Comment by valerian-chen** [comment]
I'll upload the latest version before our chat on Monday, but I'm making this comment before I forget. The issue I'm facing now is that the first point in the run always has zero density, even if it's way inside the flux surface. Very confused.

**Comment by valerian-chen** [comment]
I've fixed this in the DensityFit class

```    def __call__(self, poloidal_flux: ArrayLike) -> ArrayLike:
        """Returns the interpolated density at ``poloidal_flux`` points."""
        tol = 1e-10
        
        poloidal_flux = np.asfarray(poloidal_flux)
        density = np.asfarray(self._fit_impl(poloidal_flux))
        # Mask density inside plasma
        is_inside = poloidal_flux <= self.poloidal_flux_enter + tol
        return is_inside * density
```

**Comment by valerian-chen** [comment]
I think there's still a small error in how K_plasma is calculated from K_vacuum, as H is not exactly zero (off by about 1 percent). Happens even when the mismatch angle is zero along the entire path.

**Comment by valerian-chen** [comment]
Note to self:

If Psi_BC_flag
	set to continuous, then print deprecation warning

wrapper functions

pip install -e .

**Comment by valerian-chen** [comment]
@ZedThree I've made all the changes I'd like for the first draft. I'll be adding more test functions in the future, but I want this to run so that collaborators can start using it.

**Comment by valerian-chen** [comment]
The tests are failing because the right sign of delta_Z is not automatically chosen, so the finite differences are going out of rather than in to the plasma, at the edge.

**Comment by valerian-chen** [comment]
Somewhere, the plasma frequency receives a negative electron density and this makes some of the gradients NaN, which trips the matrix inversion up, leading to a bunch of failed tests

**Comment by valerian-chen** [comment]
I've been testing using Scotty_main_circularFluxSurfaces.py. The issue is that there's a small machine error, Scotty thinks that the density at the entry point is -7.105427357601002e-14. The negative sign gives NaNs in the Hamiltonian, then messing everything up.

@ZedThree What should we do? I'm thinking
- in fun_general, find_normalised_plasma_freq, I should round up to zero
- In the above function, add a warning if the density is negative and large in magnitude.

File: scotty/density_fit.py
Comment (ZedThree) [inline scotty/density_fit.py:43]: I think we should back out this change. Because most of the fits are defined to go to zero exactly at `poloidal_flux_enter`, this gives rise to a negative density when `poloidal_flux_enter < poloidal_flux < poloidal_flux_enter + tol`

Comment (valerian-chen) [reply scotty/density_fit.py:43]: Perhaps in the discontinuous case, we can have electron_density_plasma to be evaluated at poloidal_flux_enter-tol

File: tests/test_fun_general.py
Comment (ZedThree) [inline tests/test_fun_general.py:269]: There's some floating point error, so we want:

```suggestion
    npt.assert_allclose(Psi_3D_plasma, expected_Psi_3D_plasma)
```

**Comment by valerian-chen** [comment]
I think I finally understand how test_simple_golden works. It's such an elegant way to test many cases

File: Preprocessing/ne_write_radialcoord.py
Comment (ZedThree) [inline Preprocessing/ne_write_radialcoord.py:40]: I really recommend using the `with` syntax for working with files:

```suggestion
with open('ne.dat','w') as ne_data_file:
    ne_data_file.write(f"{int(output_length)}\n") 
    for ii in range(0, output_length):
        ne_data_file.write(f'{radialcoord_array[ii]:.8e} {density_array[ii]:.8e} \n')        
```

**Review by ZedThree — DISMISSED** [review]
It would be good to also fix the values of `Psi_BC_flag` in `get_parameters_for_scotty` to the new values

**Comment by valerian-chen** [comment]
Hey @ZedThree, I can't figure out why the tests are failing, I'm not sure where in test_initial_parameters are the types checked.

If everything else looks good, would you mind approving the review? I know it's not strictly necessary, but I thought it'd be good, just to be sure.

I've implemented your suggested changes. Not sure why it says I've dismissed your review.

**Review by ZedThree — APPROVED** [review]

**Comment by ZedThree** [comment]
Ah, the tests were failing because there's now a string in the parameters dict, and numpy doesn't like to do `allclose` with strings. I've pushed a fix:

```diff
modified   tests/test_initial_parameters.py
@@ -81,8 +81,8 @@ def test_parameters_DBS_NSTX_MAST():
     }
     assert sorted(expected.keys()) == sorted(parameters.keys())
     for key, value in expected.items():
-        if isinstance(value, (bool, type(None))):
-            assert parameters[key] is value, key
+        if isinstance(value, (bool, type(None), str)):
+            assert parameters[key] == value, key
         else:
             assert np.allclose(parameters[key], value), key
```

Don't worry about the review dismissal, Github automatically does it when changes are pushed to a PR. I've reapproved -- I'll let you merge it

**Pull request merged**
